src/conn/conn_layered.c

```diff
@@ -592,64 +530,152 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
 
     /* Add the terminating zero byte to the end of the buffer. */
     len = item.size + 1;
-    WT_ERR(__wt_calloc_def(session, len, &buf)); /* This already zeroes out the buffer. */
+    WT_ERR(__wt_calloc_def(session, len, &buf));
     memcpy(buf, item.data, item.size);
 
-    __disagg_construct_meta_config_array(buf, cfg_meta_array);
-    WT_ERR(__wt_config_getones(session, cfg_meta_array[TIMESTAMP_CONFIG], "timestamp", &cval));
-    if (cval.len > 0 && cval.val == 0)
-        checkpoint_timestamp = WT_TS_NONE;
-    else
-        WT_ERR(
-          __wt_txn_parse_timestamp(session, "checkpoint timestamp", &checkpoint_timestamp, &cval));
+    *bufp = buf;
+    buf = NULL;
 
-    /* Save the metadata key-value pair. */
-    metadata_key = WT_DISAGG_METADATA_URI;
-    metadata_value = root = buf;
+err:
+    __wt_free(session, buf);
+    __wt_free(session, item.mem);
+    return (ret);
+}
 
-    __wt_verbose_debug2(session, WT_VERB_DISAGGREGATED_STORAGE,
-      "Picking up disaggregated storage checkpoint: metadata_lsn=%" PRIu64 ", timestamp=%" PRIu64
-      " %s"
-      ", root=\"%s\"",
-      ckpt_meta->metadata_lsn, checkpoint_timestamp,
-      __wt_timestamp_to_string(checkpoint_timestamp, ts_string), root);
+/*
+ * __disagg_load_checkpoint_meta --
+ *     Parse checkpoint metadata pulled from the shared metadata buffer.
+ */
+static int
+__disagg_load_checkpoint_meta(
+  WT_SESSION_IMPL *session, char *meta_buf, char **metadata_valuep, uint64_t *checkpoint_timestampp)
+{
+    WT_CONFIG meta_cfg;
+    WT_CONFIG_ITEM cfg_key, cfg_value;
+    WT_DECL_RET;
+    uint64_t checkpoint_timestamp;
+    char *metadata_value;
+    bool checkpoint_found, timestamp_found;
 
-    /* We need an internal session when modifying metadata. */
-    WT_ERR(__wt_open_internal_session(conn, "checkpoint-pick-up", false, 0, 0, &internal_session));
+    WT_CLEAR(meta_cfg);
+    metadata_value = NULL;
+    checkpoint_timestamp = WT_TS_NONE;
+    checkpoint_found = timestamp_found = false;
+
+    __wt_config_init(session, &meta_cfg, meta_buf);
+    while ((ret = __wt_config_next(&meta_cfg, &cfg_key, &cfg_value)) == 0) {
+        __wt_verbose_debug2(session, WT_VERB_DISAGGREGATED_STORAGE,
+          "Disaggregated checkpoint metadata item \"%.*s\"=\"%.*s\"", (int)cfg_key.len, cfg_key.str,
+          (int)cfg_value.len, cfg_value.str);
+        if (WT_CONFIG_LIT_MATCH("checkpoint", cfg_key)) {
+            if (checkpoint_found)
```

Comment:
This should feel more like an assert here. If we find a duplicate checkpoint that screams trouble. We can just check if metadata_value is null rather than having checkpoint_found